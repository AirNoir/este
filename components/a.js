// @flow
import Link from 'next/link';
import Text, { type TextProps } from './text';

type AProps = TextProps & {
  href: string,
  isActive?: boolean,
  prefetch?: boolean,
};

const A = ({ href, prefetch, ...props }: AProps) => (
  <Link href={href} prefetch={prefetch}>
    <Text
      as="a"
      href={href}
      style={(theme, mixStyles) => {
        const {
          color = 'primary',
          isActive,
          rawStyle: propsRawStyle,
          ...textProps
        } = mixStyles(props);
        return {
          color,
          decoration: isActive ? 'underline' : 'none',
          rawStyle: {
            ':hover': { textDecoration: 'underline' },
            ...propsRawStyle,
          },
          ...textProps,
        };
      }}
    />
  </Link>
);

export default A;
