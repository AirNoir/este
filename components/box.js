// @flow
import type { Color, Theme } from '../themes/types';
import PropTypes from 'prop-types';
import React from 'react';
import { reject, isNil, map } from 'ramda';

/*
  Box is thin abstraction for universal, styled, typed, and themed components.
    Box - Container
    Box - Header
    Box - Text
    Box - Text - Heading
    Box - Text - Input

  Box picks props to handle, and pass the rest. No white/black listing.
    <Box
      width="10%" // width is an universal style prop
      onTouch={onBoxTouch} // onClick in a browser
      onUnsupportedProp={...} // Box will pass prop to underlying component.
    />

  Some Box styles are restricted to or leverage theme constants.
    <Box
      backgroundColor="primary" // theme.colors.primary
      padding={1} // theme.typography.rhythm(n)
    />
    <Container style={theme => ({ padding: theme.container.padding })} />

  Box supports any platform styles via rawStyle to bypass type checking.
    <Box rawStyle={theme => ({ whateverFelaOrReactNativeSupport: ... })} />
    <Box style={theme => ({
      margin: 1, // style prop
      rawStyle: whateverFelaOrReactNativeSupport
    })}/>

  We can style any component via as property.
    <Box as={Slider} {...props} />

  Box enforces vertical rhythm with borders via padding compensation.

  Box styles follow https://facebook.github.io/yoga capabilities. For browser,
  we restrict styles to enforce the same behaviour.

  For minimal builds, we recommend to use box.js, box.ios.js, etc.

  Flow Theme type should be the same for all platforms to help maintenance.

  For an inspiration:
    https://github.com/react-native-community/react-native-elements
    http://jxnblk.com/rebass
    https://vuetifyjs.com
    https://github.com/airyland/vux
    https://material-ui.com

  TODO:
    - transform: https://microsoft.github.io/reactxp/docs/styles.html
    - maybe, handle View in Text https://github.com/Microsoft/reactxp/blob/762abbe7450501fc6b1088d55ef5539dd51ff223/src/web/utils/restyleForInlineText.tsx
*/

type MaybeRhythmProp = number | string;

export type BoxProps = {
  as?: string | ((props: Object) => React.Element<*>),
  isReactNative?: boolean,
  style?: (theme: Theme, mixStyle: (Object) => Object) => BoxProps,
  rawStyle?: Object,

  margin?: MaybeRhythmProp,
  marginHorizontal?: MaybeRhythmProp,
  marginVertical?: MaybeRhythmProp,
  marginBottom?: MaybeRhythmProp,
  marginLeft?: MaybeRhythmProp,
  marginRight?: MaybeRhythmProp,
  marginTop?: MaybeRhythmProp,
  padding?: MaybeRhythmProp,
  paddingHorizontal?: MaybeRhythmProp,
  paddingVertical?: MaybeRhythmProp,
  paddingBottom?: MaybeRhythmProp,
  paddingLeft?: MaybeRhythmProp,
  paddingRight?: MaybeRhythmProp,
  paddingTop?: MaybeRhythmProp,
  height?: MaybeRhythmProp,
  minHeight?: MaybeRhythmProp,
  maxHeight?: MaybeRhythmProp,
  width?: MaybeRhythmProp,
  minWidth?: MaybeRhythmProp,
  maxWidth?: MaybeRhythmProp,
  bottom?: MaybeRhythmProp,
  left?: MaybeRhythmProp,
  right?: MaybeRhythmProp,
  top?: MaybeRhythmProp,

  // Flexbox. Only what's compatible with React Native.
  // github.com/facebook/react-native/blob/master/Libraries/StyleSheet/LayoutPropTypes.js
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline',
  alignSelf?:
    | 'auto'
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'stretch'
    | 'baseline',
  flex?: number,
  flexBasis?: number | string,
  flexDirection?: 'row' | 'row-reverse' | 'column' | 'column-reverse',
  flexGrow?: number,
  flexShrink?: number,
  flexWrap?: 'wrap' | 'nowrap',
  justifyContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around',

  backgroundColor?: Color,
  opacity?: number,
  overflow?: 'visible' | 'hidden' | 'scroll',
  position?: 'absolute' | 'relative',
  zIndex?: number,
  borderStyle?: 'solid' | 'dotted' | 'dashed',

  borderWidth?: number,
  borderBottomWidth?: number,
  borderLeftWidth?: number,
  borderRightWidth?: number,
  borderTopWidth?: number,

  borderRadius?: number,
  borderBottomLeftRadius?: number,
  borderBottomRightRadius?: number,
  borderTopLeftRadius?: number,
  borderTopRightRadius?: number,

  borderColor?: Color,
  borderBottomColor?: Color,
  borderLeftColor?: Color,
  borderRightColor?: Color,
  borderTopColor?: Color,
};

type BoxContext = {
  renderer: { renderRule: () => Object },
  theme: Theme,
};

type Transformations = {
  [prop: string]: (
    theme: Theme,
    props: BoxProps,
    prop: $Keys<BoxProps>,
    value: any
  ) => {| style?: Object, props?: Object |},
};

const maybeRhythm = theme => value =>
  typeof value === 'number' ? theme.typography.rhythm(value) : value;

const createShorthandTransformation = destructure => (options, props) => ({
  style: reject(isNil, destructure(props)),
});

const createMaybeRhythmShorthandTransformation = destructure => (
  theme,
  props
) => ({
  style: map(
    maybeRhythm(theme),
    createShorthandTransformation(destructure)(theme, props).style
  ),
});

const marginTransformation = createMaybeRhythmShorthandTransformation(
  ({
    margin,
    marginHorizontal = margin,
    marginVertical = margin,
    marginBottom = marginVertical,
    marginLeft = marginHorizontal,
    marginRight = marginHorizontal,
    marginTop = marginVertical,
  }) => ({ marginBottom, marginLeft, marginRight, marginTop })
);

const paddingTransformation = createMaybeRhythmShorthandTransformation(
  ({
    padding,
    paddingHorizontal = padding,
    paddingVertical = padding,
    paddingBottom = paddingVertical,
    paddingLeft = paddingHorizontal,
    paddingRight = paddingHorizontal,
    paddingTop = paddingVertical,
  }) => ({ paddingBottom, paddingLeft, paddingRight, paddingTop })
);

const createColorShorthandTransformation = destructure => (theme, props) => ({
  style: map(
    value => theme.colors[value],
    createShorthandTransformation(destructure)(theme, props).style
  ),
});

const maybeRhythmTransformation = (theme, props, prop, value) => ({
  style: { [prop]: maybeRhythm(theme)(value) },
});

const justValueTransformation = (options, props, prop, value) => ({
  style: { [prop]: value },
});

// https://github.com/necolas/react-native-web expandStyle-test.js
const flexTransformation = (theme, props, prop, value) => {
  if (value < 1) throw new Error('Not implemented yet');
  const { flexBasis = 'auto', flexShrink = 1 } = props;
  return {
    style: props.isReactNative
      ? { [prop]: value }
      : { flexBasis, flexGrow: value, flexShrink },
  };
};

const colorTransformation = (theme, props, prop, value) => ({
  style: { [prop]: theme.colors[value] },
});

const borderWidthTransformation = createShorthandTransformation(
  ({
    borderWidth = 0, // Enfore React Native behaviour. It also makes more sense.
    borderBottomWidth = borderWidth,
    borderLeftWidth = borderWidth,
    borderRightWidth = borderWidth,
    borderTopWidth = borderWidth,
  }) => ({
    borderBottomWidth,
    borderLeftWidth,
    borderRightWidth,
    borderTopWidth,
  })
);

const borderRadiusTransformation = createShorthandTransformation(
  ({
    borderRadius,
    borderBottomLeftRadius = borderRadius,
    borderBottomRightRadius = borderRadius,
    borderTopLeftRadius = borderRadius,
    borderTopRightRadius = borderRadius,
  }) => ({
    borderBottomLeftRadius,
    borderBottomRightRadius,
    borderTopLeftRadius,
    borderTopRightRadius,
  })
);

const borderColorTransformation = createColorShorthandTransformation(
  ({
    borderColor,
    borderBottomColor = borderColor,
    borderLeftColor = borderColor,
    borderRightColor = borderColor,
    borderTopColor = borderColor,
  }) => ({
    borderBottomColor,
    borderLeftColor,
    borderRightColor,
    borderTopColor,
  })
);

const transformations: Transformations = {
  margin: marginTransformation,
  marginHorizontal: marginTransformation,
  marginVertical: marginTransformation,
  marginBottom: marginTransformation,
  marginLeft: marginTransformation,
  marginRight: marginTransformation,
  marginTop: marginTransformation,
  padding: paddingTransformation,
  paddingHorizontal: paddingTransformation,
  paddingVertical: paddingTransformation,
  paddingBottom: paddingTransformation,
  paddingLeft: paddingTransformation,
  paddingRight: paddingTransformation,
  paddingTop: paddingTransformation,

  height: maybeRhythmTransformation,
  minHeight: maybeRhythmTransformation,
  maxHeight: maybeRhythmTransformation,
  width: maybeRhythmTransformation,
  minWidth: maybeRhythmTransformation,
  maxWidth: maybeRhythmTransformation,
  bottom: maybeRhythmTransformation,
  left: maybeRhythmTransformation,
  right: maybeRhythmTransformation,
  top: maybeRhythmTransformation,

  alignItems: justValueTransformation,
  alignSelf: justValueTransformation,
  flex: flexTransformation,
  flexBasis: justValueTransformation,
  flexDirection: justValueTransformation,
  flexGrow: justValueTransformation,
  flexShrink: justValueTransformation,
  flexWrap: justValueTransformation,
  justifyContent: justValueTransformation,

  backgroundColor: colorTransformation,
  opacity: justValueTransformation,
  overflow: justValueTransformation,
  position: justValueTransformation,
  zIndex: justValueTransformation,
  borderStyle: justValueTransformation,

  borderWidth: borderWidthTransformation,
  borderBottomWidth: borderWidthTransformation,
  borderLeftWidth: borderWidthTransformation,
  borderRightWidth: borderWidthTransformation,
  borderTopWidth: borderWidthTransformation,

  borderRadius: borderRadiusTransformation,
  borderBottomLeftRadius: borderRadiusTransformation,
  borderBottomRightRadius: borderRadiusTransformation,
  borderTopLeftRadius: borderRadiusTransformation,
  borderTopRightRadius: borderRadiusTransformation,

  borderColor: borderColorTransformation,
  borderBottomColor: borderColorTransformation,
  borderLeftColor: borderColorTransformation,
  borderRightColor: borderColorTransformation,
  borderTopColor: borderColorTransformation,
};

export const createMixStyles = (theme: Theme) => (props: Object) => ({
  ...props,
  ...(typeof props.style === 'function'
    ? props.style(theme, createMixStyles(theme))
    : null),
});

// Border breaks rhythm because it adds additional space.
// We compensate it via padding automatically, if possible.
const tryEnsureRhythmViaPaddingCompensation = style =>
  ['Bottom', 'Left', 'Right', 'Top'].reduce((style, prop) => {
    const borderXWidth = style[`border${prop}Width`];
    const paddingProp = `padding${prop}`;
    const paddingX = style[paddingProp];
    const computableNumbers =
      typeof borderXWidth === 'number' && typeof paddingX === 'number';
    if (!computableNumbers) return style;
    const compensatedPaddingX = paddingX - borderXWidth;
    if (compensatedPaddingX < 0) return style;
    return { ...style, [paddingProp]: compensatedPaddingX };
  }, style);

// Enforce the same styles for all platforms via React Native emulation.
// https://facebook.github.io/yoga
// https://github.com/Microsoft/reactxp
// https://github.com/necolas/react-native-web
const browserReactNativeEmulation = {
  display: 'flex',
  flexDirection: 'column',
  position: 'relative',
};

const computeBoxStyleAndProps = (theme: Theme, boxProps: BoxProps) => {
  let style = boxProps.isReactNative ? {} : browserReactNativeEmulation;
  let props = {};
  Object.keys(boxProps).forEach(prop => {
    const value = boxProps[prop];
    const transformation = transformations[prop];
    if (!transformation) {
      props = { ...props, [prop]: value };
      return;
    }
    // TODO: Skip already processed shorthands for better performance?
    // if (transformation === marginTransformation) etc.
    const transformed = transformation(theme, boxProps, prop, value);
    if (transformed.style) style = { ...style, ...transformed.style };
    if (transformed.props) props = { ...props, ...transformed.props };
  });
  return { style: tryEnsureRhythmViaPaddingCompensation(style), props };
};

const Box = (props: BoxProps, { renderer, theme }: BoxContext) => {
  const { as, style, rawStyle, ...restProps } = createMixStyles(theme)(props);
  const computed = computeBoxStyleAndProps(theme, restProps);
  const className = renderer.renderRule(() => ({
    ...computed.style,
    ...rawStyle,
  }));
  return React.createElement(as || 'div', { ...computed.props, className });
};

Box.contextTypes = {
  renderer: PropTypes.object,
  theme: PropTypes.object,
};

export default Box;
