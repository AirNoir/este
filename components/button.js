// @flow
import Text, { type TextProps } from './text';
import type { ColorProps } from '../themes/types';

// Use a div instead of a button because a div has no default weird styles.
// https://www.w3.org/TR/wai-aria-practices-1.1/#button
// https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_button_role#Keyboard_and_focus
const BrowserDivButton = ({
  disabled,
  onPress,
  style,
  ...restProps
}: {
  disabled?: boolean,
  onPress?: EventHandler, // eslint-disable-line no-undef
  style?: Object,
}) => (
  <div // eslint-disable-line jsx-a11y/no-static-element-interactions
    role="button"
    disabled={disabled}
    onClick={onPress}
    onKeyPress={e => {
      if (disabled) return;
      const activation = e.key === 'Enter' || e.key === ' ';
      if (!activation || !onPress) return;
      e.preventDefault();
      onPress(e);
    }}
    style={{
      ...style,
      cursor: 'pointer',
      pointerEvents: disabled ? 'none' : 'auto',
      userSelect: 'none',
    }}
    tabIndex={disabled ? -1 : 0}
    {...restProps}
  />
);

type ButtonProps = ColorProps &
  TextProps & {
    disabled?: boolean,
    onPress?: EventHandler, // eslint-disable-line no-undef
    outline?: boolean,
  };

const Button = (props: ButtonProps) => (
  <Text
    style={theme => {
      const {
        as = BrowserDivButton,
        size = 0,
        borderRadius = theme.button.borderRadius,
        marginHorizontal = theme.button.marginHorizontal,
        marginVertical = size < 0
          ? theme.button.marginVertical + theme.button.paddingVertical
          : theme.button.marginVertical,
        paddingHorizontal = 1,
        // For size < 0, these is no space for padding.
        paddingVertical = size < 0 ? 0 : theme.button.paddingVertical,
        outline,
        ...restProps
      } = props;

      // <Button primary etc. shorthand. It overrides defaults.
      const colorName = Object.keys(theme.colors).find(color => props[color]);
      if (colorName) {
        delete restProps[colorName];
        if (outline) {
          restProps.borderColor = colorName;
          restProps.borderStyle = 'solid';
          restProps.borderWidth = theme.button.borderWidth;
          restProps.color = colorName;
        } else {
          restProps.backgroundColor = colorName;
          restProps.bold = true;
          restProps.color = 'white';
        }
      }

      // Ensure vertical rhythm for small button with border. The lineHeight
      // is the only possible way how to do it. It doesn't work for multilines.
      if (size < 0 && restProps.borderWidth) {
        restProps.lineHeight =
          theme.typography.lineHeight - 2 * restProps.borderWidth;
      }

      if (restProps.disabled) {
        restProps.opacity = theme.button.disabledOpacity;
      }

      return {
        as,
        size,
        borderRadius,
        marginHorizontal,
        marginVertical,
        paddingHorizontal,
        paddingVertical,
        ...restProps,
      };
    }}
  />
);

export default Button;
