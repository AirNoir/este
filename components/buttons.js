// // @flow
// import Box, { type BoxProps } from '../components/box';
//
// // Horizontal container for buttons.
//
// export type ButtonsProps = BoxProps & {
//   vertical?: boolean,
// };
//
// const Buttons = (props: ButtonsProps) => (
//   <Box
//     style={theme => {
//       const {
//         vertical,
//         flexDirection = vertical ? 'column' : 'row',
//         flexWrap = 'wrap',
//         marginBottom = 1,
//         marginHorizontal = -theme.button.marginHorizontal,
//         ...restProps
//       } = props;
//       return {
//         flexDirection,
//         flexWrap,
//         marginBottom,
//         marginHorizontal,
//         ...restProps,
//       };
//     }}
//   />
// );
//
// export default Buttons;
