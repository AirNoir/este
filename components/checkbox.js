// // @flow
// import Button, { type ButtonProps } from './button';
// import SvgIcon from './svg-icon';
// import Text from './text';
//
// export type CheckboxProps = ButtonProps & {
//   label: string,
//   labelOnLeft?: boolean,
//   onChange?: ({ value: boolean }) => mixed,
//   value: boolean,
// };
//
// const Checkbox = ({
//   label,
//   labelOnLeft = false,
//   onChange,
//   value,
//   color,
//   size,
//   ...props
// }: CheckboxProps) => (
//   <Button
//     aria-checked={value}
//     onPress={() => onChange && onChange({ value: !value })}
//     role="checkbox"
//     style={(theme, mixStyles) => {
//       const {
//         marginVertical = 0,
//         paddingHorizontal = 0,
//         paddingVertical = 0,
//         ...restProps
//       } = mixStyles(props);
//       return {
//         ...restProps,
//         marginVertical,
//         paddingHorizontal,
//         paddingVertical,
//       };
//     }}
//   >
//     {labelOnLeft && <Text color={color} size={size}>{label}</Text>}
//     <SvgIcon
//       style={theme => ({
//         color,
//         size,
//         svg: value ? theme.checkbox.checkedIcon : theme.checkbox.uncheckedIcon,
//         [labelOnLeft ? 'marginLeft' : 'marginRight']: 0.5,
//       })}
//     />
//     {!labelOnLeft && <Text color={color} size={size}>{label}</Text>}
//   </Button>
// );
//
// export default Checkbox;
