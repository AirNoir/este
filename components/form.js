// @flow
import Box, { type BoxProps } from './box';

type FormProps = BoxProps & {
  onSubmit?: EventHandler,
};

const onSubmitWithPreventDefault = onSubmit => event => {
  event.preventDefault();
  if (!onSubmit) return;
  onSubmit(event);
};

const Form = ({ onSubmit, ...props }: FormProps) => (
  <Box
    as={'form'}
    onSubmit={onSubmitWithPreventDefault(onSubmit)}
    style={(theme, mixStyles) => {
      const {
        marginBottom = theme.form.marginBottom,
        maxWidth = theme.form.maxWidth,
        ...restProps
      } = mixStyles(props);
      return { marginBottom, maxWidth, ...restProps };
    }}
  />
);

export default Form;
