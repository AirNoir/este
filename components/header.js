// @flow
import Box from './box';
// import Link from 'next/link';
// import A from './a';

import Text from './text';

// {
//   /* <Link prefetch href="/"><A>home</A></Link>
// <Link prefetch href="/about"><A>about</A></Link> */
// }

const HeaderLink = ({ children }) => (
  <Text
    backgroundColor="primary"
    bold
    color="white"
    paddingHorizontal={0.5}
    paddingVertical={0.5}
  >
    {children}
  </Text>
);

const Header = () => (
  <Box
    backgroundColor="primary"
    flexDirection="row"
    flexWrap="wrap"
    marginVertical={0.5}
    paddingHorizontal={0.5}
  >
    <HeaderLink>Home</HeaderLink>
    <HeaderLink>About</HeaderLink>
  </Box>
);

export default Header;
