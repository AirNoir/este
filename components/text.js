// @flow
import type { Color } from '../themes/types';
import Box, { type BoxProps } from './box';
import colorLib from 'color';
import withTheme, { type ThemeContext } from './withTheme';

/*
  Text is the basic UI primitive for all text components.
    Text -> Heading
    Text -> Button
    Text -> TextInput
    etc.
*/

export type TextProps = BoxProps & {
  align?: 'left' | 'right' | 'center' | 'justify',
  bold?: boolean,
  color?: Color,
  decoration?: 'none' | 'underline' | 'line-through',
  fontFamily?: string,
  italic?: boolean,
  lineHeight?: number,
  size?: number,
  // TODO: shadowColor, shadowOffset, shadowRadius.
};

// http://inlehmansterms.net/2014/06/09/groove-to-a-vertical-rhythm
const computeFontSizeAndLineHeight = (typography, size) => {
  const fontSize = typography.fontSize(size);
  const lines = Math.ceil(fontSize / typography.lineHeight);
  const lineHeight = lines * typography.lineHeight;
  return { fontSize, lineHeight };
};

// http://usabilitypost.com/2012/11/05/stop-fixing-font-smoothing
// tldr; Fix font smoothing only for light text on a dark background.
const fixBrowserFontSmoothing = (color, backgroundColor) => {
  const hasColorAndBackgroundColor =
    color !== 'transparent' && backgroundColor !== 'transparent';
  if (!hasColorAndBackgroundColor) return null;
  const colorIsLighterThanBackgroundColor =
    colorLib(color).luminosity() > colorLib(backgroundColor).luminosity();
  if (!colorIsLighterThanBackgroundColor) return null;
  return {
    MozOsxFontSmoothing: 'grayscale',
    WebkitFontSmoothing: 'antialiased',
  };
};

const emulateReactNative = (theme, style, backgroundColor) => ({
  ...style,
  ...(backgroundColor
    ? fixBrowserFontSmoothing(style.color, theme.colors[backgroundColor])
    : null),
  lineHeight: `${style.lineHeight}px`, // browser needs px
});

const Text = (props: TextProps, { theme }: ThemeContext) => {
  const {
    align,
    bold,
    color = theme.text.color,
    decoration,
    fontFamily = theme.text.fontFamily,
    italic,
    lineHeight,
    size = 0,
    ...restProps
  } = props;

  let style = {
    color: theme.colors[color],
    display: 'inline', // React Native default Text display value.
    // https://github.com/Microsoft/reactxp/blob/328a54affdd573aa99b348e5b60e65e3d4ba57a3/src/web/Text.tsx#L24
    whiteSpace: 'pre-wrap',
    overflowWrap: 'break-word',
    msHyphens: 'auto',
    fontFamily,
    ...computeFontSizeAndLineHeight(theme.typography, size),
    ...(align ? { textAlign: align } : null),
    ...(bold ? { fontWeight: theme.text.bold } : null),
    ...(decoration ? { textDecoration: decoration } : null),
    ...(italic ? { fontStyle: 'italic' } : null),
    ...(lineHeight ? { lineHeight } : null),
    ...restProps.style,
  };

  style = emulateReactNative(theme, style, restProps.backgroundColor);

  return <Box {...restProps} style={style} />;
};

withTheme(Text);

export default Text;
