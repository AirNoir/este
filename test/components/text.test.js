// @flow
import { computeStyle } from '../../components/text';

const browserDefaultStyle = {
  // Enforce React Native behaviour for browsers.
  // display: 'flex',
  // flexDirection: 'column',
  // position: 'relative',
  // overflow: 'hidden',
};

const theme = {
  typography: {
    fontSize: level => level * 2 + 16, // 0 = 16, 1 = 18, -1 = 14 etc.
    rhythm: ratio => ratio * 24,
    lineHeight: 24,
  },
  colors: {
    primary: 'blue',
    success: 'green',
    warning: 'orange',
    danger: 'red',
    black: 'black',
    white: 'white',
    gray: 'gray',
  },
  text: {
    bold: 600,
    color: 'black',
    // github.com/twbs/bootstrap/blob/v4-dev/scss/_variables.scss
    fontFamily: '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif',
    rawStyle: {
      fixFontSmoothing: true,
    },
  },
};

// $ FlowFixMe Don't fix. We test real values, not types.
const compute = props => computeStyle(props, { isReactNative: true })(theme);

// $ FlowFixMe Don't fix. We test real values, not types.
const computeBrowser = props =>
  computeStyle(props, { isReactNative: false })(theme);

test('text without props', () => {
  const boxProps = compute({});
  expect(boxProps).toMatchSnapshot();
});

test('browser text without props', () => {
  const boxProps = computeBrowser({});
  expect(boxProps).toMatchSnapshot();
});

test('props', () => {
  expect(compute({ align: 'left' })).toMatchSnapshot();
  expect(compute({ bold: true })).toMatchSnapshot();
  expect(compute({ color: 'primary' })).toMatchSnapshot();
  expect(compute({ decoration: 'underline' })).toMatchSnapshot();
  expect(compute({ fontFamily: 'arial' })).toMatchSnapshot();
  expect(compute({ italic: true })).toMatchSnapshot();
  expect(compute({ lineHeight: 10 })).toMatchSnapshot();
  expect(compute({ size: 1 })).toMatchSnapshot();
});

test('fix fontSmoothing for light text on a dark background', () => {
  expect(
    computeBrowser({ color: 'primary', backgroundColor: 'black' })
  ).toMatchSnapshot();
});

test('do not fix fontSmoothing for dark text on a light background', () => {
  expect(
    computeBrowser({ color: 'primary', backgroundColor: 'white' })
  ).toMatchSnapshot();
});

test('do not fix fontSmoothing for dark text without backgroundColor', () => {
  expect(computeBrowser({ color: 'primary' })).toMatchSnapshot();
});

test('lineHeight ensures vertical rhythm', () => {
  // $FlowFixMe
  expect(compute({ size: 4 }).rawStyle.lineHeight).toBe(
    theme.typography.lineHeight * 1
  );
  // $FlowFixMe
  expect(compute({ size: 5 }).rawStyle.lineHeight).toBe(
    theme.typography.lineHeight * 2
  );
  // $FlowFixMe
  expect(compute({ size: 16 }).rawStyle.lineHeight).toBe(
    theme.typography.lineHeight * 2
  );
  // $FlowFixMe
  expect(compute({ size: 17 }).rawStyle.lineHeight).toBe(
    theme.typography.lineHeight * 3
  );
});
