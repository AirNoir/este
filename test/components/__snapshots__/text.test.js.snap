// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`browser text without props 1`] = `
Object {
  "rawStyle": Object {
    "color": "black",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "lineHeight": "24px",
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`do not fix fontSmoothing for dark text on a light background 1`] = `
Object {
  "backgroundColor": "white",
  "rawStyle": Object {
    "color": "blue",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "lineHeight": "24px",
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`do not fix fontSmoothing for dark text without backgroundColor 1`] = `
Object {
  "rawStyle": Object {
    "color": "blue",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "lineHeight": "24px",
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`fix fontSmoothing for light text on a dark background 1`] = `
Object {
  "backgroundColor": "black",
  "rawStyle": Object {
    "MozOsxFontSmoothing": "grayscale",
    "WebkitFontSmoothing": "antialiased",
    "color": "blue",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "lineHeight": "24px",
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`props 1`] = `
Object {
  "rawStyle": Object {
    "color": "black",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "lineHeight": 24,
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "textAlign": "left",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`props 2`] = `
Object {
  "rawStyle": Object {
    "color": "black",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "fontWeight": 600,
    "lineHeight": 24,
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`props 3`] = `
Object {
  "rawStyle": Object {
    "color": "blue",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "lineHeight": 24,
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`props 4`] = `
Object {
  "rawStyle": Object {
    "color": "black",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "lineHeight": 24,
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "textDecoration": "underline",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`props 5`] = `
Object {
  "rawStyle": Object {
    "color": "black",
    "display": "inline",
    "fontFamily": "arial",
    "fontSize": 16,
    "lineHeight": 24,
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`props 6`] = `
Object {
  "rawStyle": Object {
    "color": "black",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "fontStyle": "italic",
    "lineHeight": 24,
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`props 7`] = `
Object {
  "rawStyle": Object {
    "color": "black",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "lineHeight": 10,
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`props 8`] = `
Object {
  "rawStyle": Object {
    "color": "black",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 18,
    "lineHeight": 24,
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;

exports[`text without props 1`] = `
Object {
  "rawStyle": Object {
    "color": "black",
    "display": "inline",
    "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, sans-serif",
    "fontSize": 16,
    "lineHeight": 24,
    "msHyphens": "auto",
    "overflowWrap": "break-word",
    "whiteSpace": "pre-wrap",
  },
}
`;
